File: src/world/world.ts
import NanoTimer from 'nanotimer'
import { actionVectors } from '../actionVectors'
import { Agent } from '../entities/agent/agent'
import { Player } from '../entities/agent/player'
import { Blade } from '../entities/blade'
import { Circle } from '../entities/circle'
import { Door } from '../entities/door'
import { Star } from '../entities/star'
import { Transporter } from '../entities/transporter'
import { add, clamp, clampVec, combine, dirFromTo, dot, getDistance, mul, normalize, range, sub, X, Y } from '../math'
import { Rock } from '../entities/rock'

export class World {
  timer = new NanoTimer()
  rocks: Rock[] = []
  agents: Agent[] = []
  players: Player[] = []
  blades: Blade[] = []
  stars: Star[] = []
  circles: Circle[] = []
  doors: Door[] = []
  walls: number[][][] = []
  boundary: number[][] = []
  transporters: Transporter[] = []
  timeStep = 0.02
  timeScale = 1
  time = 0
  busy = false
  paused = false

  constructor() {
    this.timer.setInterval(() => this.step(), '', `${this.timeStep / this.timeScale}s`)
  }

  addPlayer(position: number[]): Player {
    const player = new Player(this, position)
    return player
  }

  addAgent(position: number[]): Agent {
    const agent = new Agent(this, position)
    return agent
  }

  addRock(position: number[], radius: number): Rock {
    const rock = new Rock(this, position, radius)
    return rock
  }

  addPlayerBlade(position: number[]): Blade {
    const blade = new Blade(this, position)
    blade.align = 0
    return blade
  }

  addGuardBlade(position: number[]): Blade {
    const blade = new Blade(this, position)
    blade.align = 1
    return blade
  }

  addDoor(vector: number[], shell: number[][]): Door {
    const door = new Door(this, vector, shell)
    return door
  }

  addStar(position: number[]): Star {
    const star = new Star(this, position)
    return star
  }

  addTransporter(position: number[], target: number[]): Transporter {
    const transporter = new Transporter(this, position, target)
    return transporter
  }

  preStep(): void { }
  postStep(): void { }

  step(): void {
    if (this.busy) {
      console.log('busy')
      return
    }
    if (this.paused) return
    this.busy = true
    this.preStep()
    const dt = this.timeStep
    this.time += dt
    this.circles.forEach(circle => {
      circle.force = [0, 0]
      circle.impulse = [0, 0]
      circle.shift = [0, 0]
    })
    this.agents.forEach(agent => {
      if (agent.dead) return
      agent.force = mul(agent.movePower, actionVectors[agent.action])
    })
    this.blades.forEach(blade => {
      if (blade.agent != null) {
        const vector = sub(blade.agent.position, blade.position)
        const clampedVector = clampVec(vector, 10)
        blade.force = mul(blade.movePower, clampedVector)
      }
    })
    this.doors.forEach(door => { door.onStep(dt) })
    this.transporters.forEach(transporter => {
      let chargeRate = -1
      this.players.forEach(player => {
        const distance = getDistance(player.position, transporter.center)
        if (distance > 1) return
        if (transporter.charge === transporter.interval) {
          transporter.transport(player)
          return
        }
        chargeRate = 1
      })
      transporter.charge = clamp(0, transporter.interval, transporter.charge + chargeRate * dt)
    })
    this.players.forEach(player => {
      this.stars.forEach(star => {
        if (star.agent != null) return
        if (star.door != null) return
        if (player.star != null) return
        const distance = getDistance(player.position, star.spawnPoint)
        if (distance > Star.radius + Agent.radius) return
        player.takeStar(star)
      })
      if (player.dead) {
        player.respawn()
      }
    })
    this.agents.forEach(agent => {
      if (agent.dead) return
      this.checkBlades(agent)
    })
    this.agents.forEach(agent1 => {
      if (agent1.dead) return
      this.agents.forEach(agent2 => {
        if (agent2.dead) return
        if (agent1.index >= agent2.index) return
        this.collideCircleCircle(agent1, agent2)
      })
    })
    this.rocks.forEach(rock => {
      this.agents.forEach(agent => {
        if (agent.dead) return
        this.collideCircleCircle(rock, agent)
      })
      this.rocks.forEach(rock2 => {
        this.collideCircleCircle(rock, rock2)
      })
    })
    this.blades.forEach(blade1 => {
      this.blades.forEach(blade2 => {
        if (blade1.index >= blade2.index) return
        this.collideCircleCircle(blade1, blade2)
      })
    })
    this.circles.forEach(circle => {
      this.collideCircleWall(circle, this.boundary)
      this.walls.forEach(wall => {
        this.collideCircleWall(circle, wall)
      })
      this.doors.forEach(door => {
        const hit = this.collideCircleWall(circle, door.shell)
        if (hit && circle instanceof Player) door.knock(circle)
      })
    })
    this.circles.forEach(circle => {
      circle.velocity = mul(1 - circle.drag * dt, circle.velocity)
      circle.velocity = combine(1, circle.velocity, dt / circle.mass, circle.force)
      circle.velocity = combine(1, circle.velocity, 1 / circle.mass, circle.impulse)
      circle.position = combine(1, circle.position, dt, circle.velocity)
      circle.position = combine(1, circle.position, 1, circle.shift)
    })
    this.circles.forEach(circle => {
      circle.history.unshift(circle.position)
      circle.history = circle.history.slice(0, Circle.historyLength)
    })
    this.postStep()
    this.busy = false
  }

  collideCircleCircle(circle1: Circle, circle2: Circle): boolean {
    const distance = getDistance(circle1.position, circle2.position)
    const overlap = circle1.radius + circle2.radius - distance
    if (overlap <= 0) return false
    const normal = dirFromTo(circle1.position, circle2.position)
    const relativeVelocity = sub(circle1.velocity, circle2.velocity)
    const impactSpeed = dot(relativeVelocity, normal)
    const massFactor = 1 / (1 / circle1.mass + 1 / circle2.mass)
    const impulse = mul(impactSpeed * massFactor, normal)
    const shift = mul(0.5 * overlap, normal)
    circle1.impulse = combine(1, circle1.impulse, -1, impulse)
    circle2.impulse = combine(1, circle2.impulse, +1, impulse)
    circle1.shift = combine(1, circle1.shift, -1, shift)
    circle2.shift = combine(1, circle2.shift, +1, shift)
    return true
  }

  collideCircleWall(circle: Circle, wall: number[][]): boolean {
    let hit = false
    for (const i of range(wall.length)) {
      const j = i > 0 ? i - 1 : wall.length - 1
      const a = wall[i]
      const b = wall[j]
      const c = circle.position
      const ab = sub(b, a)
      const ac = sub(c, a)
      const bc = sub(c, b)
      const dir = normalize(ab)
      const normal = [-dir[Y], +dir[X]]
      if (dot(normal, ac) < 0) {
        normal[X] = -normal[X]
        normal[Y] = -normal[Y]
      }
      if (dot(ac, ab) < 0) continue
      if (dot(bc, ab) > 0) continue
      const overlap = circle.radius - dot(ac, normal)
      if (overlap < 0) continue
      const impactSpeed = -dot(circle.velocity, normal)
      const impulse = mul(1.2 * impactSpeed * circle.mass, normal)
      circle.impulse = add(circle.impulse, impulse)
      const shift = mul(overlap, normal)
      circle.shift = add(circle.shift, shift)
      hit = true
    }
    if (hit) return hit
    for (const point of wall) {
      if (this.collideCirclePoint(circle, point)) {
        hit = true
      }
    }
    return hit
  }

  collideCirclePoint(circle: Circle, point: number[]): boolean {
    const distance = getDistance(circle.position, point)
    const overlap = circle.radius - distance
    if (overlap <= 0) return false
    const normal = dirFromTo(point, circle.position)
    const impactSpeed = -dot(circle.velocity, normal)
    const impulse = mul(1.2 * impactSpeed * circle.mass, normal)
    circle.impulse = add(circle.impulse, impulse)
    const shift = mul(overlap, normal)
    circle.shift = add(circle.shift, shift)
    return true
  }

  checkBlades(agent: Agent): void {
    if (agent.dead) return
    this.blades.forEach(blade => {
      if (agent.dead) return
      const distance = getDistance(agent.position, blade.position)
      const overlap = agent.radius + blade.radius - distance
      if (overlap < 0) return
      if (blade.align === agent.align) {
        if (agent.blade == null && blade.agent == null) blade.attach(agent)
        return
      }
      agent.die()
      return
    })
  }
}

File: model/world.py
from __future__ import annotations
from math import cos, pi, sin
import torch
import torch.nn.functional as F

# device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
# defaultType = torch.float64
torch.set_printoptions(sci_mode=False)

# actionVectorList = [[0.0,0.0]]
# for i in range(8):
#     angle = 2 * pi * i / 8
#     dir = [cos(angle), sin(angle)]
#     actionVectorList.append(dir)
# actionVectors = torch.tensor(actionVectorList).to(device)
# actions = torch.tensor([i for i in range(9)]).to(device)
# actionIndexPairs = torch.cartesian_prod(actions,actions)
# actionVectorPairs = torch.cat(
#     [
#         actionVectors[actionIndexPairs[:,0]],
#         actionVectors[actionIndexPairs[:,1]]
#     ],
#     dim=1
# )

class Circle: 
    def __init__(self,world: World, radius: float):
        self.world = world
        self.radius = radius
        self.mass = pi * radius ** 2
        self.drag = 0
        self.position = torch.zeros(world.count,2,dtype=world.dtype).to(world.device)
        self.velocity = torch.zeros(world.count,2,dtype=world.dtype).to(world.device)
        self.force = torch.zeros(world.count,2,dtype=world.dtype).to(world.device)
        self.impulse = torch.zeros(world.count,2,dtype=world.dtype).to(world.device)
        self.shift = torch.zeros(world.count,2,dtype=world.dtype).to(world.device)
        self.id = len(self.world.circles)
        self.world.circles.append(self)

class Agent(Circle):
    def __init__(self, world: World):
        super().__init__(world, 0.5)
        self.blade = Blade(self)
        self.movePower = 3
        self.drag = 0.7
        self.dead = torch.zeros(world.count, dtype=torch.int).to(world.device)
        self.action = torch.zeros(world.count, dtype=torch.int).to(world.device)
        self.world.agents.append(self)

class Blade(Circle):
    def __init__(self, agent: Agent):
        super().__init__(agent.world, 1)
        self.agent = agent
        self.movePower = 4
        self.drag = 0.3
        self.world.blades.append(self)

class World:
    def __init__(self, count: int, device: torch.device, dtype = torch.float64):
        self.timeStep = 0.04
        self.circles: list[Circle] = []
        self.agents: list[Agent] = []
        self.blades: list[Blade] = []
        self.count = count
        self.device = device
        self.dtype = dtype
        actionVectorList = [[0.0,0.0]]
        for i in range(8):
            angle = 2 * pi * i / 8
            dir = [cos(angle), sin(angle)]
            actionVectorList.append(dir)
        self.actionVectors = torch.tensor(actionVectorList).to(device)
        self.actions = torch.tensor([i for i in range(9)]).to(device)

    def checkDeath(self, agent: Agent):
        for blade in self.blades:
            if blade.agent.id == agent.id: continue
            vector = agent.position - blade.position
            distance = torch.linalg.norm(vector, dim=1)
            overlap = agent.radius + blade.radius - distance
            agent.dead = torch.where(overlap > 0, 1, agent.dead)

    def collideCircleCircle(self, circle1: Circle, circle2: Circle):
        if circle1.id >= circle2.id: return
        vector = circle2.position - circle1.position
        distance = torch.linalg.norm(vector, dim=1)
        overlap = (circle1.radius + circle2.radius - distance).unsqueeze(1)
        normal = F.normalize(vector, dim=1)
        relativeVelocity = circle1.velocity - circle2.velocity
        impactSpeed = torch.linalg.vecdot(relativeVelocity, normal).unsqueeze(1)
        massFactor = 1 / circle1.mass + 1 / circle2.mass
        impulse = torch.where(overlap > 0, impactSpeed / massFactor * normal, 0)
        shift = torch.where(overlap > 0, 0.5 * overlap * normal, 0)
        circle1.impulse = circle1.impulse - impulse
        circle2.impulse = circle2.impulse + impulse
        circle1.shift = circle1.shift - shift
        circle2.shift = circle2.shift + shift
            
    def step(self):
        dt = self.timeStep
        agentCount = len(self.agents)
        for i in range(agentCount):
            agent = self.agents[i]
            blade = self.blades[i]
            agent.force[:,:] = 0
            blade.force[:,:] = 0
            agent.impulse[:,:] = 0
            blade.impulse[:,:] = 0
            agent.shift[:,:] = 0
            blade.shift[:,:] = 0
        for agent in self.agents:
            agent.force = agent.movePower * self.actionVectors[agent.action]
        for blade in self.blades:
            vector = blade.agent.position - blade.position
            blade.force = blade.movePower * vector
        for agent1 in self.agents:
            for agent2 in self.agents:
                self.collideCircleCircle(agent1, agent2)
        for blade1 in self.blades:
            for blade2 in self.blades:
                self.collideCircleCircle(blade1, blade2)
        for circle in self.circles:
            circle.velocity = (1 - circle.drag * dt) * circle.velocity 
            circle.velocity = circle.velocity + dt / circle.mass * circle.force
            circle.velocity = circle.velocity + 1 / circle.mass * circle.impulse
            circle.position = circle.position + dt * circle.velocity + circle.shift
        for agent in self.agents:
            self.checkDeath(agent)
        for agent in self.agents:
            direction = F.normalize(agent.position, dim=1)
            dead = (agent.dead > 0).unsqueeze(1)
            agent.position = torch.where(dead, agent.position + 15 * direction, agent.position)
            agent.blade.position = torch.where(dead, agent.position.clone(), agent.blade.position)
            agent.velocity =  torch.where(dead, 0*agent.velocity, agent.velocity)
            agent.blade.velocity = torch.where(dead, 0*agent.blade.velocity, agent.blade.velocity)
            agent.dead = 0 * agent.dead




